### PR TITLE
Migrate `dependencies.yml` to Gradle's version catalogs

### LINKED ISSUE #4120: Use version catalogs for dependency management
Version catalog is now a stable feature since Gradle 7.4. It's time to migrate from our in-house `dependencies.yaml`. See https://docs.gradle.org/7.4/userguide/platforms.html for more information.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

We can declare dependency versions using `dependencies.yml` and share them
across multiple projects. It is useful to manage the versions of various
dependencies. However, we can't define different versions of a module.
Because the unique keys in `dependencies.yml` are `moduleId:artifactId`.

This limitation makes it difficult to manage compatibility between
different versions of a library. For example, okhttp v4 is required to
test `it:okhttp` and okhttp v3 is required to run `retrofit2` module.

Gradle introduced a new way to [share dependency versions between projects](https://docs.gradle.org/current/userguide/platforms.html).
The unique key of version catalogs is a user-defined alias so we can
define multiple versions of a module using different aliases.

Version catalogs do not allow extra properties on `libs.versions.toml`.
It only allows some properties for versions and modules.
`dependencies.yml` not only manages versions but also manages metadata
for a module such as relocations, exclusions and Javadoc links.

I wanted to take advantage of both. We need a version manager that
supports version catalogs based on aliases, and the ability to add the
metadata. As a result, I implemented to add our own dependency mechanism
using `dependencies.toml` that is fully compatible with Gradle's version
catalogs and understands relocations, exclusions and Javadoc links
declarations.

Modifications:

- Migrate `dependencies.yml` to `dependencies.toml`
- Rename `common-dependences.gradle` to `common-dependences-lagacy.gradle` 
  for backward compatibility with `dependencies.yml`
  - Newly added `common-dependences.gradle` handles the dependencies declared by `dependencies.toml`
- Add `version-catalog.gradle` that parses `dependencies.toml` and builds:
  - Library and BOM version catalogs
  - relocations
  - exclusions
  - Javadoc links
- Migrate all dependencies defined in `build.gradle` files to the new version catalog API.

Result:

Fixes #4120

### CODE REVIEW

**Review by minwoox — COMMENTED** [review]
It's way simpler! Looks great! 😄

File: gradle/scripts/lib/java-alpn.gradle
Comment (ikhoon) [reply gradle/scripts/lib/java-alpn.gradle:17]: There are two problems with `libs...`. 
- We don't assume that the version of Jetty ALPN is defined at `libs.versions.jetty.alpn.agent`.
- It could be a breaking change for the users who do not use `dependencies.toml`.

File: gradle/scripts/lib/scala.gradle
Comment (ikhoon) [reply gradle/scripts/lib/scala.gradle:62]: > We are going to eventually remove `managedVersions`, right?

I initially thought so but now I don't think we can remove `managedVersions`. Because we have to enforce users to use the alias defined at `gradle-scripts`.
It is convenient for users to use any name without strong naming conventions.

File: annotation-processor/build.gradle
Comment (ikhoon) [reply annotation-processor/build.gradle:3]: Nope. It should be removed.

File: gradle/scripts/lib/scala.gradle
Comment (minwoox) [reply gradle/scripts/lib/scala.gradle:62]: That makes sense. 😄 
After I read [this](https://github.com/line/armeria/pull/4353/files#diff-d984c6b3a5597bcee9d1c32ba3261f2bdfbef1d2d7b0c4aa68686ff5fdda9c21R196), I thought we are going to remove that eventually, but it's not. 😄

File: build.gradle
Comment (ikhoon) [reply build.gradle:19]: Without these lines, Gradle raises an exception when applying the plugin.
For example, if `alias libs.plugins.kotlin apply false` is removed:
```
Script '/Users/ikhun/src/armeria/gradle/scripts/lib/kotlin.gradle' line: 3

* What went wrong:
A problem occurred evaluating script.
> Plugin with id 'kotlin' not found.

```

Comment (minwoox) [reply build.gradle:19]: Ah ha, that is correct. Thanks!

File: gradle/scripts/version-catalogs.gradle
Comment (ikhoon) [reply gradle/scripts/version-catalogs.gradle:50]: 😱😱😱

File: dependencies.toml
Comment (ikhoon) [reply dependencies.toml:58]: Let me remove it

File: gradle/scripts/lib/scala.gradle
Comment (ikhoon) [reply gradle/scripts/lib/scala.gradle:62]: `managedVersions` has a known limitation but we need to use it until we find a better answer. 😅

File: gradle/scripts/README.md
Comment (jrhee17) [inline gradle/scripts/README.md:198]: I see, so it is possible that the version returned by `managedVersions` isn't necessarily equal to the eventually resolved version. Am I understanding correctly?

**Review by jrhee17 — COMMENTED** [review]
Mostly looks good! Only left a few questions 👍

File: build.gradle
Comment (ikhoon) [reply build.gradle:315]: This is useful to debug some dependencies. Let me move it to `version-catalogs.gradle`

File: dependencies.toml
Comment (ikhoon) [reply dependencies.toml:62]: > I think I personally prefer defining dependencies with a . delimiter.

You meant `akka.grpc.runtime.v213`? TOML interpret `.` into a nested table like:
```
akka: {
    "http" : {
      "cors" : {
         ...
```

File: dependencies.toml
Comment (ikhoon) [reply dependencies.toml:1024]: It should be `0.6.6` 😅

File: dependencies.toml
Comment (jrhee17) [reply dependencies.toml:62]: Ah I see 😅  Ignore this comment please

File: thrift0.16/build.gradle
Comment (ikhoon) [reply thrift0.16/build.gradle:9]: Previously, it was possible because one of the multiple versions was set as `constraints`.
But now `thrift0.13` must not use `libthrift:0.16`. Anyway, it makes sense to set `strictly` to the thrift versions in `dependencies.toml`.

File: gradle/scripts/README.md
Comment (ikhoon) [reply gradle/scripts/README.md:198]: `managedVersions` can be safely used  when a `groupId:artifactId` is declared only once in a whole `dependencies.toml`.

Let's say the following dependencies are defined.
```toml
[libraries.jsoup]
module = "org.jsoup:jsoup"
version = "1.15.1"

[libraries.thrift09]
module = "org.apache.thrift:libthrift"
version = "0.9.3-1"
[libraries.thrift012]
module = "org.apache.thrift:libthrift"
version = "0.12.0"
```

We can assume that `managedVersions['org.jsoup:jsoup']` returns `"1.15.1"`. However, we can't assert whether `managedVersions['org.jsoup:jsoup']` returns `"0.9.3-1"` or `"0.12.0"`.

**Review by jrhee17 — APPROVED** [review]
I'm assuming the generated artifacts are the same before/after the change.  Looks good to me 👍

File: dependencies.toml
Comment (minwoox) [inline dependencies.toml:48]: So the convention of specifying version is that
- the version is inlined if it's used only once.
- the version is specified in `[versions]` section if it's used more than once.
Is that correct?

**Review by minwoox — APPROVED** [review]
The build script is much simplified! Thanks, @ikhoon!

**Comment by minwoox** [comment]
We have to also check if https://github.com/line/armeria-examples/blob/master/update-examples.sh is working correctly before merging this. 😉

File: dependencies.toml
Comment (ikhoon) [reply dependencies.toml:48]: Correct. But we can freely change the convention anytime. 😆

**Comment by ikhoon** [comment]
> We have to also check if [https://github.com/line/armeria-examples/blob/master/update-examples.sh](https://github.com/line/armeria-examples/blob/master/update-examples.sh?rgh-link-date=2022-07-29T08%3A00%3A40Z) is working correctly before merging this. 😉

`update-examples.sh` should not work at all. I'd like to fix it in the f/u PR before we release a new version.

**Comment by ikhoon** [comment]
Let me merge this PR without waiting for CI builds because the last commit only changed `gradle/scritps/README.md`.

**Pull request merged**
